README.md
# Matrix

1. 类似[Python](https://www.python.org)这样的开发语言， 在使用过程中，很可能使用到第三方扩展依赖库，这些依赖库需要一个C语言编译环境，但是编译工具往往比较庞大，
例如GCC的编译工具链就比较庞大，如果直接在Docker基础镜像中包含[GCC](https://gcc.gnu.org/)编译工具链，就会构造的应用程序镜像的大小也会随之增加。
2. Python提供了[PIP](https://pip.pypa.io/en/stable/)这样包管理工具，也能很好解决递归依赖的问题，但是他需要一个[PYPI](https://pypi.python.org/pypi)私有仓库。
如果这个私有仓库的设置散落在众多应用程序的Dockerfile中，管理起来也是比较麻烦，特别是私有仓库迁移，可能对应用程序来说更是要命。  


这就是matrix产生原因，是构建Python项目变得容易，简单。

## Usage

如果要使用matrix来构建镜像，项目根目录中需要包含一个matrix.json元数据描述文件， 来告诉matrix如何构建镜像。可以查看matrix-schema.json文件来了解matrix.json
的详细定义。

### matrix-schema.json

```json

{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "image": {
      "type": "string"
    },
    "tag": {
      "type": "string"
    },
    "port": {
      "type": "array",
      "items": {
        "type": "integer"
      }
    },
    "env": {
      "type": "object"
    },
    "volume": {
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "cmd": {
      "type": "array"
    }
  },
  "required": [
    "name",
    "image",
    "tag"
  ]
}

```

使用也非常简单， 以matrix镜像启动一个容器， 把你的Python项目的路径映射容器中/home/matrix路径，并且/var/run/docker.sock 也映射到容器的/var/run/docker.sock，
执行/usr/local/bin/matrix.sh, 执行命令如下：

```shell
sudo docker run --rm -v /home/benjamin/git/otter:/home/matrix -v $(which docker):$(which docker) -v /var/run/docker.sock:/var/run/docker.sock matrix:0.0.2 /usr/local/bin/matrix.sh

```

## Exit Status

1. 80 - /home/matrix is not exists
2. 81 - install requirements.txt error
3. 82 - run custom build.sh error
4. 83 - build image error
5. 84 - matrix.json is not exists
6. 85 - matrix.json content invalid, read the matrix-schema.json for detail

## 部署结构
构建成功之后，代码会部署到容器中/opt/biz目录，并且当前工作目录也会被设置成/opt/biz


HISTORY.md
Matrix
===================

0.0.1
----------------------------
1. add [alpine build tools chain](http://pkgs.alpinelinux.org/package/v3.4/main/x86_64/build-base)
2. add git
3. add python 2.7.3, python-dev
4. add [freetds](http://www.freetds.org), freetds-dev
5. add docker

0.0.2
--------------------------
1. add [virtualenv](https://virtualenv.pypa.io/en/stable)
2. add [Fabric](http://www.fabfile.org)
3. add [jsonschema](https://github.com/Julian/jsonschema)
4. validate matrix.json

fabfile.py
#! /usr/bin/env python
import argparse
import fnmatch
import json
import locale
import logging
import os
import re
import shutil
import sys
from contextlib import contextmanager
from distutils.text_file import TextFile

import jsonschema

INCLUDE = []
INCLUDE_regexps = []
EXCLUDE = []
EXCLUDE_regexps = []


# -------------------------
# copy file
# --------------------------
@contextmanager
def cd(directory):
    """Change the current working directory, temporarily.
    Use as a context manager: with cd(d): ...
    """
    old_dir = os.getcwd()
    try:
        os.chdir(directory)
        yield
    finally:
        os.chdir(old_dir)


def get_all_file_in(dirname):
    name_list = []
    for root, dirs, files in os.walk(dirname):
        # for dir in dirs:
        #     if isinstance(dir,bytes):
        #         name_list.append(dir.decode(locale.getpreferredencoding()))
        #     else:
        #         name_list.append(dir)
        for name in files:
            if isinstance(name, bytes):
                name_list.append(os.path.join(root, name.decode(locale.getpreferredencoding())))
            else:
                name_list.append(os.path.join(root, name))
    return name_list


def _glob_to_regexp(pat):
    pat = fnmatch.translate(pat)
    sep = r'\\\\' if os.path.sep == '\\' else os.path.sep
    return re.sub(r'((?<!\\)(\\\\)*)\.', r'\1[^%s]' % sep, pat)


def read_manifest():
    if not os.path.isfile('MANIFEST.in'):
        return
    include, include_regexps, exclude, exclude_regexps = _get_message_from_manifest('MANIFEST.in')
    INCLUDE.extend(include)
    INCLUDE_regexps.extend(include_regexps)
    EXCLUDE.extend(exclude)
    EXCLUDE_regexps.extend(exclude_regexps)


def _get_message_from_list(list):
    include, include_regexps, exclude, exclude_regexps = _get_message_from_manifest_lines(list)
    INCLUDE.extend(include)
    INCLUDE_regexps.extend(include_regexps)
    EXCLUDE.extend(exclude)
    EXCLUDE_regexps.extend(exclude_regexps)


def _get_message_from_manifest(filename):
    template = TextFile(filename,
                        strip_comments=True,
                        skip_blanks=True,
                        join_lines=True,
                        lstrip_ws=True,
                        rstrip_ws=True,
                        collapse_join=True)
    try:
        lines = template.readlines()
    finally:
        template.close()
    return _get_message_from_manifest_lines(lines)


def _get_message_from_manifest_lines(lines):
    follow = []
    follow_regexps = []
    ignore = []
    ignore_regexps = []
    for line in lines:
        try:
            cmd, rest = line.split(None, 1)
        except ValueError:
            continue
        if cmd == 'exclude':
            for pat in rest.split():
                if '*' in pat or '?' in pat or '[!' in pat:
                    ignore_regexps.append(_glob_to_regexp(pat))
                else:
                    ignore.append(pat)
        elif cmd == 'global-exclude':
            ignore.extend(rest.split())
        elif cmd == 'recursive-exclude':
            try:
                dirname, patterns = rest.split(None, 1)
            except ValueError:
                continue
            dirname = dirname.rstrip(os.path.sep)
            for pattern in patterns.split():
                if pattern.startswith('*'):
                    ignore.append(dirname + os.path.sep + pattern)
                else:
                    ignore.append(dirname + os.path.sep + pattern)
                    ignore.append(dirname + os.path.sep + '*' + os.path.sep +
                                  pattern)
        elif cmd == 'prune':
            rest = rest.rstrip('/\\')
            ignore.append(rest)
            ignore.append(rest + os.path.sep + '*')
        elif cmd == 'include':
            for pat in rest.split():
                if '*' in pat or '?' in pat or '[!' in pat:
                    follow_regexps.append(_glob_to_regexp(pat))
                else:
                    follow.append(pat)
        elif cmd == 'global-include':
            follow.extend(rest.split())
        elif cmd == 'recursive-include':
            try:
                dirname, patterns = rest.split(None, 1)
            except ValueError:
                continue
            dirname = dirname.rstrip(os.path.sep)
            for pattern in patterns.split():
                if pattern.startswith('*'):
                    follow.append(dirname + os.path.sep + pattern)
                else:
                    follow.append(dirname + os.path.sep + pattern)
                    follow.append(dirname + os.path.sep + '*' + os.path.sep +
                                  pattern)
        elif cmd == 'graft':
            rest = rest.rstrip('/\\')
            follow.append(rest)
            follow.append(rest + os.path.sep + '*')
    return follow, follow_regexps, ignore, ignore_regexps


def file_matches(filename, patterns):
    """Does this filename match any of the patterns?"""
    return any(fnmatch.fnmatch(filename, pat) or
               fnmatch.fnmatch(os.path.basename(filename), pat)
               for pat in patterns)


def file_matches_regexps(filename, patterns):
    """Does this filename match any of the regular expressions?"""
    return any(re.match(pat, filename) for pat in patterns)


def strip_sdist_extras(filelist):
    """Strip generated files that are only present in source distributions.

    We also strip files that are ignored for other reasons, like
    command line arguments, setup.cfg rules or MANIFEST.in rules.
    """
    # file_names = os.path.
    return [name for name in filelist
            if (not file_matches(os.path.basename(name), EXCLUDE)
                and not file_matches_regexps(os.path.basename(name), EXCLUDE_regexps)
                and (file_matches(os.path.basename(name), INCLUDE)
                     or file_matches_regexps(os.path.basename(name), INCLUDE_regexps)))]


def copy_files(srcdir, destdir, rule=None):
    with cd(srcdir):
        if isinstance(rule, list):
            _get_message_from_list(rule)
        elif rule is None:
            read_manifest()
    all_source_file = get_all_file_in(srcdir)
    source_file = strip_sdist_extras(all_source_file)
    for filepath in source_file:
        tmp_path = os.path.relpath(filepath, srcdir)
        destfile = os.path.join(destdir, tmp_path)
        # filename should not be absolute, but let's double-check
        # assert destfile.startswith(destdir + os.path.sep)
        destfiledir = os.path.dirname(destfile)
        if not os.path.isdir(destfiledir):
            os.makedirs(destfiledir)
        # if os.path.isdir(filename):
        #     os.mkdir(destfile)
        # else:
        shutil.copy2(filepath, destfile)


def load_settings(src):
    full_path = os.path.join(src, 'matrix.json')
    with open(full_path, 'rb') as f:
        try:
            return json.load(f)
        except ValueError as e:
            logging.error("%s is invalid json file: %s", full_path, e)
            sys.exit(1)


def gen_docker_file(src='.', matrix='.'):
    settings = load_settings(src)
    cmd = settings.get('cmd', [])
    settings['cmd'] = ','.join('"%s"' % x for x in cmd)

    port = settings.get('port', [])
    settings['expose'] = ''
    if port:
        settings['expose'] = 'EXPOSE ' + ' '.join(str(x) for x in port)

    volume = settings.get('volume', [])
    settings['volume'] = ''
    if volume:
        settings['volume'] = 'VOLUME [ ' + ', '.join('"%s"' % v for v in volume) + ' ]'

    env = settings.get('env', {})
    settings['env'] = '\r\n'.join(['ENV {0} {1}'.format(*item) for item in env.iteritems()])

    docker_file_template = os.path.join(matrix, 'Dockerfile.in')
    with open(docker_file_template, 'rb') as f:
        template = f.read()
        docker_file = os.path.join(src, 'Dockerfile')
        with open(docker_file, 'wb') as out:
            out.write(template.format(**settings))


def image_name(src='.'):
    settings = load_settings(src)
    print '{name}:{tag}'.format(name=settings.get('name'), tag=settings.get('tag', 'latest')),


def valid_matrix_json(src='.', matrix='.'):
    with open(os.path.join(src, 'matrix.json'), 'rb') as meta:
        meta_json = json.load(meta)
        with open(os.path.join(matrix, 'matrix-schema.json'), 'rb') as schema:
            schema_json = json.load(schema)
            jsonschema.validate(meta_json, schema_json)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Generate Dockerfile from source tree",
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--src', dest='src', help='The source code path')
    opts = parser.parse_args(sys.argv[1:])
    gen_docker_file(opts.src)


matrix.sh
#!/usr/bin/env sh

MATRIX_HOME=$(cd `dirname $0`; pwd)

WORKDIR=/home/matrix
TMP=/tmp/images
DEP=/tmp/matrix

alias fab="fab --fabfile=${MATRIX_HOME}/fabfile.py"

# check volume directory exists
[ ! -d ${WORKDIR} ] && exit 80

# check matrix.json exists
[ ! -f "${WORKDIR}/matrix.json" ] && exit 84

# check matrix.json validate
fab valid_matrix_json:src=${WORKDIR},matrix=${MATRIX_HOME}
[ $? -gt 0 ] && exit 85


cd ${WORKDIR}

# virtualenv ${DEP}

source ${DEP}/bin/activate

if [ -f requirements.txt ]; then
	pip install --trusted-host 10.16.78.86 -i http://10.16.78.86:3141/simple -r requirements.txt
	# check install result
	[ $? -gt 0 ] && exit 81
fi

if [ -f build.sh ]; then
	chmod +x build.sh
	./build.sh
	# check run custom build.sh exit status
	[ $? -gt 0 ] && exit 82
fi

mkdir ${TMP}

# collect source file
mkdir -p ${TMP}/opt
cp  -rvf ${WORKDIR} ${TMP}/opt/biz

# collect dependency libs
mkdir -p ${TMP}/usr/lib/python2.7/
cp -rvf ${DEP}/lib/python2.7/site-packages ${TMP}/usr/lib/python2.7/

# package resource
tar zcf images.tgz -C ${TMP}  .

# generate Dockerfile
fab gen_docker_file:matrix=${MATRIX_HOME}

# build images
docker build -t $(fab image_name | head -n 1) .

# check docker build result
[ $? -gt 0 ] && exit 83

exit 0

